examples/lib/stories/collision_detection/quadtree_example.dart

```diff
@@ -0,0 +1,443 @@
+  }
+}
+
+//#endregion
+
+//#region Environment
+
+class Brick extends SpriteComponent
+    with CollisionCallbacks, GameCollideable, UpdateOnce {
+  Brick({
+    required super.position,
+    required super.size,
+    required super.priority,
+    required super.sprite,
+  }) {
+    initCenter();
+    initCollision();
+  }
+
+  bool rendered = false;
+
+  @override
+  void renderTree(Canvas canvas) {
+    if (!rendered) {
+      super.renderTree(canvas);
+    }
+  }
+}
+
+class Water extends SpriteComponent
+    with CollisionCallbacks, GameCollideable, UpdateOnce {
+  Water({
+    required super.position,
+    required super.size,
+    required super.priority,
+    required super.sprite,
+  }) {
+    initCenter();
+    initCollision();
+  }
+}
+
+mixin GameCollideable on PositionComponent {
+  void initCollision() {
+    add(RectangleHitbox()..collisionType = CollisionType.passive);
+  }
+
+  void initCenter() {
+    cachedCenter =
+        Vector2(position.x + tileSize / 2, position.y + tileSize / 2);
+  }
+
+  late final Vector2 cachedCenter;
+}
+
+//#endregion
+
+//#region Utils
+
+mixin UpdateOnce on PositionComponent {
+  bool updateOnce = true;
+
+  @override
+  void updateTree(double dt) {
+    if (updateOnce) {
+      super.updateTree(dt);
+      updateOnce = false;
+    }
+  }
+}
+
+class StaticLayer extends PreRenderedLayer {
+  StaticLayer();
+
+  List<PositionComponent> components = [];
+
+  @override
+  void drawLayer() {
+    for (final element in components) {
+      if (element is Brick) {
+        element.rendered = false;
+        element.renderTree(canvas);
+        element.rendered = true;
+      }
+    }
+  }
+}
+
+class LayerComponent extends PositionComponent {
+  LayerComponent(this.layer);
+
+  StaticLayer layer;
+
+  @override
+  void render(Canvas canvas) {
+    layer.render(canvas);
+  }
+}
+
+extension Vector2Ext on Vector2 {
+  Vector2 translate(double x, double y) {
+    return Vector2(this.x + x, this.y + y);
+  }
+}
+
+class QuadTreeDebugComponent extends PositionComponent with HasPaint {
+  QuadTreeDebugComponent(QuadTreeCollisionDetection cd) {
+    dbg = QuadTreeNodeDebugInfo.init(cd);
+    paint.color = Colors.blue;
+    paint.style = PaintingStyle.stroke;
+    priority = 10;
+  }
+
+  late final QuadTreeNodeDebugInfo dbg;
+
+  @override
+  void render(Canvas canvas) {
+    final nodes = dbg.nodes;
+    // final textPaint = TextPaint(
+    //   style: const TextStyle(fontSize: 5.0, color: Colors.lightGreenAccent),
+    // );
+    for (final node in nodes) {
+      canvas.drawRect(node.rect, paint);
+      final nodeElements = node.ownElements;
+      Paint? boxPaint;
+      if (!node.noChildren && nodeElements.isNotEmpty) {
+        boxPaint = Paint();
+        boxPaint.style = PaintingStyle.stroke;
+        boxPaint.color = Colors.lightGreenAccent;
+        boxPaint.strokeWidth = 1;
+      }
+      for (final box in nodeElements) {
+        // textPaint.render(canvas, node.id.toString(), box.aabb.min);
```

Comment:
This can be removed right?

Reply:
I'm not sure, could it be useful for understanding or not: it prints element's current node ID, showing, how it changes while component moving from one node to another. 
But I commented this out, because text rendering burns out all performance optimisations: I have nearly 3 FPS with so huge map and text rendering on every Brick

Reply:
It sounds like it is too heavy to be useful for now then.